### PR TITLE
Add Redshift support

### PR DESCRIPTION
**Pull request by aterekhov-plr**
This small fix checks that the database is Redshift and make client-side quoting instead of server quoting.

### CODE REVIEW

**Comment by dossy** [comment]
I don't have a Redshift cluster handy to test this PR against, so I'm labeling this PR as `help wanted` in case someone else who has access to a Redshift cluster would like to test this.  Looking at the PR, it's unclear why this change is even necessary, as the server-side quoting should work for ordinary Postgres as well as Redshift.

@aterekhov-plr what is the error you get when trying to use dbmate with Redshift, that this PR fixes?  In other words: how can someone else verify that there's a problem that is fixed by this PR?

**Comment by gregwebs** [comment]
@dossy the error was shown here: https://github.com/amacneil/dbmate/discussions/361

**Comment by aterekhov-plr** [comment]
@dossy If you have time to test this PR and help with the release, I could try to provide access to a temporary Redshift cluster for you.

**Comment by dossy** [comment]
@aterekhov-plr That would be helpful, if you could provide temporary access.  I can find the time next week to work on this.  My email is listed on my GitHub profile, and so are [my PGP/GPG keys](https://github.com/dossy.gpg).

**Comment by amacneil** [comment]
Looking at how other software handles redshift (e.g. [golang-migrate](https://pkg.go.dev/github.com/golang-migrate/migrate/database/redshift), [goose](https://github.com/pressly/goose/blob/master/internal/dialect/dialectquery/redshift.go), official [redshift go package](https://pkg.go.dev/github.com/aws/aws-sdk-go/service/redshift)), I think it might make more sense to add an entirely new `redshift://` driver (even if it mostly inherits from the postgres driver). That way we can allow the user to be explicit rather than implicit auto-detection.

File: pkg/dbmate/version.go [hunk 1]
```diff
@@ -1,4 +1,4 @@
 package dbmate
 
 // Version of dbmate
-const Version = "2.6.0"
```
Comment (amacneil) [inline pkg/dbmate/version.go:4]: can you remove this please, we update the version only when making releases

**Comment by aterekhov-plr** [comment]
> Looking at how other software handles redshift (e.g. [golang-migrate](https://pkg.go.dev/github.com/golang-migrate/migrate/database/redshift), [goose](https://github.com/pressly/goose/blob/master/internal/dialect/dialectquery/redshift.go), official [redshift go package](https://pkg.go.dev/github.com/aws/aws-sdk-go/service/redshift)), I think it might make more sense to add an entirely new `redshift://` driver (even if it mostly inherits from the postgres driver). That way we can allow the user to be explicit rather than implicit auto-detection.

I agree that creating a new driver is the best approach to implement Redshift support. I began working on it but later realized that the difference between the Redshift driver and the PostgreSQL driver would only be a few lines of code. I might revisit it later, but I currently need some assistance with tests. There have been issues with ports (5432/5439) if use pg, and it's unclear what to use in the docker-compose setup.

**Comment by dossy** [comment]
@aterekhov-plr, @amacneil -- I started down the path of trying to create an entire separate `redshift` driver package, but as golang doesn't have OO where I could "subclass" the `postgres` driver, it resulted in mostly a duplicated copy of the code.  I opted to register `redshift` as a driver prefix, instead.

I created a variant of one of the PG tests that I could run against the Redshift instance that @aterekhov-plr provided (thanks!) so that I could demonstrate that the original driver implementation "failed" and introducing Andrei's fix makes the test pass, so I'm happy with this.

In order to run the Redshift test, set the `REDSHIFT_TEST_URL` environment variable with a `redshift://` connection string to a Redshift cluster, before running the tests.

Let me know what you both think, but I think this is good to go, unless either of you feel otherwise?

**Comment by pavel-popov-playrix** [comment]
@dossy @amacneil do you need a temporary redshift cluster for testing or can we terminate it? Cos right now we're spending extra money for it, so if you no longer need the cluster we will terminate it.

**Comment by dossy** [comment]
@pavel-popov-playrix I was able to complete my testing using it, so I'm okay with terminating it.  Thank you so much for your generosity.

File: .gitignore [hunk 2]
```diff
@@ -9,3 +9,4 @@
 /vendor
 dist
 node_modules
+docker-compose.override.yml
```
Comment (amacneil) [inline .gitignore:12]: nit: I try to keep this file alphabetically sorted

File: pkg/driver/postgres/postgres.go [hunk 3]
```diff
@@ -115,10 +124,12 @@ func (drv *Driver) openPostgresDB() (*sql.DB, error) {
...
 
-	// connect to postgres database
-	postgresURL.Path = "postgres"
+	// connect to postgres database, unless this is a Redshift connection
+	if drv.databaseURL.Scheme != "redshift" {
+		postgresURL.Path = "postgres"
+	}
 
-	return sql.Open("postgres", postgresURL.String())
+	return sql.Open("postgres", connectionString(postgresURL))
```
Comment (amacneil) [inline pkg/driver/postgres/postgres.go:132]: why is this necessary? we already ran it through `connectionString()` above

File: pkg/driver/postgres/postgres.go [hunk 4]
```diff
@@ -115,10 +124,12 @@ func (drv *Driver) openPostgresDB() (*sql.DB, error) {
 		return nil, err
 	}
 
-	// connect to postgres database
-	postgresURL.Path = "postgres"
+	// connect to postgres database, unless this is a Redshift connection
+	if drv.databaseURL.Scheme != "redshift" {
+		postgresURL.Path = "postgres"
```
Comment (amacneil) [inline pkg/driver/postgres/postgres.go:129]: If we don't change the root DB here, does that mean that `create` / `drop` / `dump` etc don't work at all?

In that case maybe it is better to give a specific error message rather than silently failing?

File: .gitignore (hunk 2 above)
Comment (amacneil) [inline .gitignore:12]: nit: I try to keep this file alphabetically sorted

**Review by amacneil — COMMENTED** [review]
Looking good, I added a few comments/questions

File: pkg/driver/postgres/postgres.go (hunk 4 above)
Comment (dossy) [reply pkg/driver/postgres/postgres.go:129]: On standard out-of-the-box Redshift, there is no database named `postgres`.

The [documentation](https://docs.aws.amazon.com/redshift/latest/dg/r_DROP_DATABASE.html) specifies that you cannot drop the `dev`, `padb_harvest`, `template0`, `template1`, or `sys:internal` databases, nor the current database.  So, for Redshift, we could try connecting to the `sys:internal` database in order to drop the target database, as long as it's not one of the listed databases.

I don't think we'll silently fail.  We should get an error from Redshift that the database cannot be dropped, most likely because it's the current database.

File: pkg/driver/postgres/postgres.go (hunk 3 above)
Comment (dossy) [reply pkg/driver/postgres/postgres.go:132]: Good catch, not sure why I thought this was necessary.

File: pkg/driver/postgres/postgres.go (hunk 4 above)
Comment (dossy) [reply pkg/driver/postgres/postgres.go:129]: @aterekhov-plr How important is it for dbmate to be able to drop/create the Redshift database?

Comment (dossy) [reply pkg/driver/postgres/postgres.go:129]: @amacneil How do you feel about this?  I like @aterekhov-plr's approach of moving forward as-is, and if anyone needs the functionality, they can open an issue and we can sort it out then?

Comment (amacneil) [reply pkg/driver/postgres/postgres.go:129]: I don't think we need drop/create functionality for redshift, as long as the error message is clear.

**Review by amacneil — APPROVED** [review]
Thank you!
